pkg/cache/nodes.go

```diff
@@ -92,62 +91,45 @@ func (nc *schedulerNodes) addExistingAllocation(pod *v1.Pod) error {
 			return err
 		}
 	}
-	return fmt.Errorf("cannot existing allocation because pod %s is not assigned to a node", pod.Name)
+	return nil
+}
+
+func (nc *schedulerNodes) addNode(node *v1.Node) {
+	nc.addAndReportNode(node, true)
 }
 
-func (nc *schedulerNodes) addNode(obj interface{}) {
+func (nc *schedulerNodes) addAndReportNode(node *v1.Node, reportNode bool) {
 	nc.lock.Lock()
 	defer nc.lock.Unlock()
 
-	node, err := convertToNode(obj)
-	if err != nil {
-		log.Logger.Error("node conversion failed", zap.Error(err))
-		return
-	}
-
-	// add node to cache
-	log.Logger.Info("adding node to cache", zap.String("NodeName", node.Name))
-	nc.cache.AddNode(node)
-
 	// add node to nodes map
-	newNode := newSchedulerNode(node.Name, string(node.UID), common.GetNodeResource(&node.Status), nc.proxy)
-	nc.nodesMap[node.Name] = newNode
+	if _, ok := nc.nodesMap[node.Name]; !ok {
+		log.Logger.Info("adding node to context",
+			zap.String("nodeName", node.Name),
+			zap.String("UID", string(node.UID)))
+		newNode := newSchedulerNode(node.Name, string(node.UID), common.GetNodeResource(&node.Status), nc.proxy)
+		nc.nodesMap[node.Name] = newNode
+	}
 
 	// once node is added to scheduler, first thing is to recover its state
-	dispatcher.Dispatch(CachedSchedulerNodeEvent{
-		NodeId:    newNode.name,
-		Event:     events.RecoverNode,
-	})
+	// node might already be in healthy state, previously recovered during recovery process,
+	// do not trigger recover again in this case.
+	if reportNode {
+		if node, ok := nc.nodesMap[node.Name]; ok {
+			if node.getNodeState() == events.States().Node.New {
+				dispatcher.Dispatch(CachedSchedulerNodeEvent{
+					NodeId: node.name,
+					Event:  events.RecoverNode,
+				})
+			}
+		}
+	}
 }
 
-func (nc *schedulerNodes) updateNode(oldObj, newObj interface{}) {
+func (nc *schedulerNodes) updateNode(oldNode, newNode *v1.Node) {
 	nc.lock.Lock()
 	defer nc.lock.Unlock()
```

Comment:
Why are we locking the schedulerNodes here?
The only thing we do is generate an event to pass on to the proxy, we're not updating the cache any more which I think removes the need for locking completely

oops read over the update call line, cancel this looks good now